### ISSUE TITLE
Integration to MobilePay Transaction Reporting API

### ISSUE THREAD

**Issue by joandrsn** [f-klubben/stregsystemet#203]
It would be awesome to integrate to the [MobilePay Transaction Reporting API](https://developer.mobilepay.dk/transactionreporting-api).

This could potentially be used for scenarios as these:
* Ping the treo on mail/slack when a new payment is ready (Maybe only twice a day to reduce stress in the treo)
* Automatically add funds to members balance

In order to be sure that each payment have been seen - the existing payments could have a transaction ID on it so that we know which payments in MobilePay that have been seen.

**Comment by cogitantium** [comment]
Yes, very good idea!

I've spoken with Forstyrrelsen about this a while ago and we must be very sure that it functions correctly if any payments are automised at all.
A django-page that tries to do this join from Mobilepay transactions CSV or the API could be a good aid for the Treo. Features such as marking a Mobilepay transaction as 'completed' and 'pending' would help in the workflow of processing payments for the Treo.

**Comment by cogitantium** [comment]
An heuristic for automatically matching payments to fembers could be to use telephone numbers to users. They're not available initially but an existing match indicates that this user has this phone number.

Only the last four digits are available from mobilepay but I believe those are random and _maybe_ uniformly distributed.

**Comment by hrjakobsen** [comment]
Don't people already write their username in the message? Matching the phone number would not allow paying for others.

**Comment by cogitantium** [comment]
Yes, absolutely. The best values to match for is the comment. I'm thinking exact match, then `longest common substring .len/ comment.len`, and lastly just fuzzy match confidence. 
If it's an exact match, we're pretty sure - even more so if name of payment matches fember user found.

**Comment by hrjakobsen** [comment]
Only exact match should be used for automatic matching. Otherwise it should be a manual task.

**Comment by cogitantium** [comment]
I should've elaborated. Matching can be done automatic but I agree that automatically submitting payments on less than exact matches is unwise.
The plan is to use the resulting confidence to suggest a user to the Treo-member handling payments, which they can then edit or sign off on. Very simplified, the UI could turn out something like the following, where the member-column acts as the existing payment-widget for fuzzily searching for a fember but in this case pre-filled with suggestion. Exact matches are pre-checked in the approval-column. 
```
Amount | Comment            | Member | Confidence | Approve
42       tester               tester   100%         [x]
42       tester (er indsat)   tester   35%          []

[Submit payments]
```
_I'm no IxD'er, suggestions welcome_

If this change is applied a migration should follow which adds a blank-allowed field to Payments for `transaction-id`. This can enable further checks which ensures that no mobilepay payment is submitted twice.

**Comment by hrjakobsen** [comment]
I'd be very hesitant to guess, even it we are fairly sure. I think we should fetch the information we are certain about, and leave the rest for a manual task. Because then if the username is in the comment, then it is simply a 5 second task of copy and pasting into the member field.

Also I think a better solution to managing duplicates is to create a new model with a transaction ID and a foreign key to a payment, instead of extending the payment model. This way payments are still a separate concern from mobile pay (as it should be imo).

**Comment by Rene4100** [comment]
Do have in mind that we did talk about this previously in the TREO, we got told by (at that time) Danske Bank that a system integration to MobilePay would increase the fee we have to pay to them by ~2-4,- kr

**Comment by cogitantium** [comment]
Good point @Rene4100, as far as I've been told by @joandrsn, it's free for less than ~100 calls a day under our current account.
However, this doesn't really apply to PR #204 as it takes a CSV from the admin web-interface of MobilePay, which is free.

**Comment by joandrsn** [comment]
@Rene4100, as @cogitantium points out, I believe the MobilePay Integration is free.
Each of the endpoint have individual pricing - but for our needs (maybe a couple of times a day) it should not be a problem.
The documentation that I'm referring to is here: https://developer.mobilepay.dk/node/3832

The only problems that I have run into is the signup process to become a developer. It's rather cumbersome.

**Comment by cogitantium** [comment]
Implemented by #204
